File: src/mvmctl/api/cache.py
"""Cache management API — orchestrates cache pruning operations.

This module provides the API layer for cache management, orchestrating
prune operations across VMs, networks, images, and kernels.
"""

from __future__ import annotations

import logging

from mvmctl.api.host import check_privileges_interactive
from mvmctl.api.metadata import get_default_image_entry, get_default_kernel_entry
from mvmctl.api.network import get_network_leases, list_networks, remove_network
from mvmctl.api.vms import remove_vm
from mvmctl.constants import DEFAULT_NETWORK_NAME, SUPPORTED_IMAGE_EXTENSIONS
from mvmctl.core import cache_manager as core_cache_manager
from mvmctl.core.metadata import (
    list_image_entries,
    list_kernel_entries,
    remove_image_entry,
    remove_kernel_entry,
)
from mvmctl.core.vm_manager import get_vm_manager
from mvmctl.models.vm import VMStatus
from mvmctl.utils.fs import (
    get_cache_dir,
    get_images_dir,
    get_kernels_dir,
)

logger = logging.getLogger(__name__)

__all__ = [
    "init_all",
    "prune_vms",
    "prune_networks",
    "prune_images",
    "prune_kernels",
    "prune_all",
]


def init_all() -> dict[str, str]:
    """Initialize all cache resources.

    Creates all necessary cache directories and initializes metadata files.
    Requires proper group privileges (checked via check_privileges).

    Returns:
        Dictionary mapping resource names to their directory paths as strings.
    """
    result = core_cache_manager.cache_init_all()
    return {k: str(v) if v else "" for k, v in result.items()}


def _get_image_references() -> set[str]:
    """Get set of image paths referenced by all VMs."""
    vm_manager = get_vm_manager()
    vms = vm_manager.list_all()

    referenced: set[str] = set()
    for vm in vms:
        if vm.config and vm.config.rootfs_path:
            referenced.add(str(vm.config.rootfs_path))

    return referenced


def _get_kernel_references() -> set[str]:
    """Get set of kernel paths referenced by all VMs."""
    vm_manager = get_vm_manager()
    vms = vm_manager.list_all()

    referenced: set[str] = set()
    for vm in vms:
        if vm.config and vm.config.kernel_path:
            referenced.add(str(vm.config.kernel_path))

    return referenced


def _get_network_references() -> set[str]:
    """Get set of network names referenced by all VMs."""
    vm_manager = get_vm_manager()
    vms = vm_manager.list_all()

    referenced: set[str] = set()
    for vm in vms:
        if vm.network_name:
            referenced.add(vm.network_name)

    return referenced


def prune_vms(
    include_stopped: bool = False,
    include_running: bool = False,
    dry_run: bool = False,
) -> list[str]:
    """Prune VMs based on their status.

    By default, only prunes VMs in ERROR state. Use flags to include
    stopped or running VMs.

    Args:
        include_stopped: Also prune STOPPED VMs.
        include_running: Also prune RUNNING VMs (use with caution).
        dry_run: If True, only report what would be removed.

    Returns:
        List of VM names that were removed.
    """
    check_privileges_interactive("/usr/sbin/ip", "prune VMs")
    vm_manager = get_vm_manager()
    vms = vm_manager.list_all()

    removed: list[str] = []
    for vm in vms:
        should_remove = False

        if vm.status == VMStatus.ERROR:
            should_remove = True
        elif vm.status == VMStatus.STOPPED and include_stopped:
            should_remove = True
        elif vm.status == VMStatus.RUNNING and include_running:
            should_remove = True

        if should_remove:
            if not dry_run:
                try:
                    remove_vm(vm.name)
                    removed.append(vm.name)
                except Exception as e:
                    logger.warning(f"Failed to remove VM {vm.name}: {e}")
            else:
                removed.append(vm.name)

    return removed


def prune_networks(dry_run: bool = False, include_all: bool = False) -> list[str]:
    """Prune unused networks.

    Args:
        dry_run: If True, only report what would be removed.
        include_all: If True, remove all networks including default and referenced.

    Returns:
        List of network names that were removed.
    """
    check_privileges_interactive("/usr/sbin/ip", "prune networks")
    referenced_networks = _get_network_references()
    all_networks = list_networks()

    removed: list[str] = []
    for network in all_networks:
        if not include_all:
            if network.name == DEFAULT_NETWORK_NAME:
                continue
            if network.name in referenced_networks:
                continue
            leases = get_network_leases(network.name)
            if leases:
                continue

        if not dry_run:
            try:
                remove_network(network.name)
                removed.append(network.name)
            except Exception as e:
                logger.warning(f"Failed to remove network {network.name}: {e}")
        else:
            removed.append(network.name)

    return removed


def prune_images(dry_run: bool = False, include_all: bool = False) -> list[str]:
    """Prune unused images.

    Args:
        dry_run: If True, only report what would be removed.
        include_all: If True, remove all images including default and referenced.

    Returns:
        List of image IDs that were removed.
    """
    cache_dir = get_cache_dir()
    images_dir = get_images_dir()

    referenced_paths = _get_image_references()
    all_images = list_image_entries(cache_dir, images_dir)

    default_entry = get_default_image_entry()
    default_id = default_entry[0] if default_entry else None

    removed: list[str] = []
    for image_id, meta in all_images.items():
        if not include_all:
            if image_id == default_id:
                continue

            path = str(meta.get("path", ""))
            if path:
                image_path = str(images_dir / path)
                if image_path in referenced_paths:
                    continue

            os_slug = str(meta.get("os_slug", ""))
            if os_slug:
                is_referenced = False
                for ref_path in referenced_paths:
                    if os_slug in ref_path:
                        is_referenced = True
                        break
                if is_referenced:
                    continue

        path = str(meta.get("path", ""))
        if not dry_run:
            try:
                if path:
                    (images_dir / path).unlink(missing_ok=True)
                else:
                    for ext in SUPPORTED_IMAGE_EXTENSIONS:
                        (images_dir / f"{image_id}{ext}").unlink(missing_ok=True)
                remove_image_entry(cache_dir, image_id)
                removed.append(image_id)
            except Exception as e:
                logger.warning(f"Failed to remove image {image_id}: {e}")
        else:
            removed.append(image_id)

    return removed


def prune_kernels(dry_run: bool = False, include_all: bool = False) -> list[str]:
    """Prune unused kernels.

    Args:
        dry_run: If True, only report what would be removed.
        include_all: If True, remove all kernels including default and referenced.

    Returns:
        List of kernel IDs that were removed.
    """
    cache_dir = get_cache_dir()
    kernels_dir = get_kernels_dir()

    referenced_paths = _get_kernel_references()
    all_kernels = list_kernel_entries(cache_dir, kernels_dir)

    default_entry = get_default_kernel_entry(cache_dir)
    default_id = default_entry[0] if default_entry else None

    removed: list[str] = []
    for kernel_id, meta in all_kernels.items():
        path = str(meta.get("path", ""))
        if not include_all:
            if kernel_id == default_id:
                continue
            if path:
                kernel_path = str(kernels_dir / path)
                if kernel_path in referenced_paths:
                    continue

        if not dry_run:
            try:
                if path:
                    (kernels_dir / path).unlink(missing_ok=True)
                else:
                    (kernels_dir / kernel_id).unlink(missing_ok=True)
                remove_kernel_entry(cache_dir, kernel_id)
                removed.append(kernel_id)
            except Exception as e:
                logger.warning(f"Failed to remove kernel {kernel_id}: {e}")
        else:
            removed.append(kernel_id)

    return removed


def prune_all(
    include_stopped: bool = False,
    include_running: bool = False,
    dry_run: bool = False,
) -> dict[str, list[str] | bool]:
    """Prune all cache resources.

    Performs a complete prune operation across all resource types:
    VMs, networks, images, and kernels.

    Args:
        include_stopped: Include stopped VMs in pruning.
        include_running: Include running VMs in pruning (use with caution).
        dry_run: If True, only report what would be removed.

    Returns:
        Dictionary with results per resource type:
        - "vms": list of removed VM names
        - "networks": list of removed network names
        - "images": list of removed image IDs
        - "kernels": list of removed kernel IDs
    """
    check_privileges_interactive("/usr/sbin/ip", "prune all cache resources")
    return {
        "vms": prune_vms(include_stopped, include_running, dry_run),
        "networks": prune_networks(dry_run),
        "images": prune_images(dry_run),
        "kernels": prune_kernels(dry_run),
    }

File: pkg/api/cache.go
// Package api provides the public orchestration layer for all operations.
// Matches src/mvmctl/api/cache_operations.py exactly.
package api

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"sort"
	"strings"

	"mvmctl/internal/infra"
	"mvmctl/internal/infra/event"
	"mvmctl/internal/lib/model"
	"mvmctl/internal/lib/provisioner/guestfs"
	"mvmctl/internal/lib/system"
	"mvmctl/pkg/api/responses"
	"mvmctl/pkg/errs"
)

// CacheCheckPrivileges checks if the current process has the required system privileges
// for destructive cache operations. Returns nil if OK, or an error describing what's missing.
func (op *Operation) CacheCheckPrivileges(binary, operation string) error {
	return system.CheckPrivileges(binary, operation)
}

// CacheSessionHasGroup returns true if the current process has the mvm group active.
func (op *Operation) CacheSessionHasGroup() bool {
	return system.SessionHasGroup()
}

// CacheInitAll initializes all cache directories.
// Matches Python's CacheOperation.init_all() exactly.
func (op *Operation) CacheInitAll(
	ctx context.Context,
	onProgress event.OnProgressCallback,
) (*responses.CacheInitResult, error) {
	cacheDir := op.CacheDir
	var created []string

	// Ensure DB schema exists before any DB writes. (Python: Database().migrate())
	if op.Connection != nil {
		if _, err := op.Connection.RunMigrationsCtx(ctx); err != nil {
			slog.Warn("Failed to run DB migrations during cache init", "error", err)
		}
	}

	// Core directories (Python: CacheUtils.get_vms_dir(), get_images_dir(), etc.)
	// Python: 6 directories: vms, images, kernels, bin, logs, keys
	// Go: Use infra.Get*Dir() functions which create and return the canonical paths.
	dirFuncs := []func() string{
		infra.GetVmsDir,
		infra.GetImagesDir,
		infra.GetKernelsDir,
		infra.GetBinDir,
		infra.GetLogsDir,
		infra.GetKeysDir,
	}

	for _, fn := range dirFuncs {
		dir := fn()
		created = append(created, dir)
	}

	// Check whether guestfs was enabled by the user
	guestfsEnabled := false
	raw, err := op.ConfigGet(ctx, "settings", "guestfs_enabled")
	if err == nil {
		if b, ok := raw.(bool); ok {
			guestfsEnabled = b
		}
	}

	// libguestfs fixed appliance (heavy operation) — only when enabled
	// Python: if guestfs_enabled: GuestfsService.build_appliance(cache_dir)
	var appliancePath string
	if guestfsEnabled {
		if onProgress != nil {
			onProgress(event.Progress{
				Phase:   "appliance",
				Status:  "running",
				Message: "Building libguestfs appliance...",
			})
		}
		appliancePath, _ = guestfs.BuildAppliance(ctx, cacheDir)
	}

	// Detected guestfs kernel (Python: KernelDetector.find_best_kernel())
	kernelPath, _, _ := guestfs.FindBestKernel(ctx)

	return &responses.CacheInitResult{
		CacheDir:         cacheDir,
		Directories:      created,
		GuestfsAppliance: appliancePath,
		GuestfsKernel:    kernelPath,
	}, nil
}

// CachePruneVMs prunes VMs via VMOperation.Prune.
// Matches Python's CacheOperation.prune_vms() exactly.
func (op *Operation) CachePruneVMs(ctx context.Context, dryRun bool, includeAll bool) *errs.OperationResult {
	ids, err := op.VMPrune(ctx, dryRun, includeAll)
	if err != nil {
		return &errs.OperationResult{
			Status:    "error",
			Code:      "cache.prune_failed",
			Message:   fmt.Sprintf("Failed to prune VMs: %v", err),
			Exception: err,
		}
	}
	return &errs.OperationResult{
		Status: "success",
		Code:   "cache.pruned",
		Item:   ids,
	}
}

// CachePruneNetworks prunes networks via NetworkOperation.Prune.
// Matches Python's CacheOperation.prune_networks() exactly.
func (op *Operation) CachePruneNetworks(ctx context.Context, dryRun bool, includeAll bool) ([]string, error) {
	return op.NetworkPrune(ctx, dryRun, includeAll)
}

// CachePruneImages prunes images via ImageOperation.Prune.
// Matches Python's CacheOperation.prune_images() exactly.
func (op *Operation) CachePruneImages(ctx context.Context, dryRun bool, includeAll bool) ([]string, error) {
	return op.ImagePrune(ctx, dryRun, includeAll)
}

// CachePruneKernels prunes kernels via KernelOperation.Prune.
// Matches Python's CacheOperation.prune_kernels() exactly.
func (op *Operation) CachePruneKernels(ctx context.Context, dryRun bool, includeAll bool) ([]string, error) {
	return op.KernelPrune(ctx, dryRun, includeAll)
}

// CachePruneBinaries prunes binaries via BinaryOperation.Prune.
// Matches Python's CacheOperation.prune_binaries() exactly.
func (op *Operation) CachePruneBinaries(ctx context.Context, dryRun bool, includeAll bool) ([]string, error) {
	return op.BinaryPrune(ctx, dryRun, includeAll)
}

// CachePruneMisc prunes miscellaneous cache items.
func (op *Operation) CachePruneMisc(ctx context.Context, dryRun bool) (map[string]any, error) {
	binDir := infra.GetBinDir()
	serviceBinariesCleaned := false
	if _, err := os.Stat(binDir); err == nil && !dryRun {
		os.RemoveAll(binDir)
		serviceBinariesCleaned = true
	}

	appliancePruned := guestfs.PruneAppliance(op.CacheDir, dryRun)
	warmPruned := op.Services.Cache.PruneWarmImages(ctx, dryRun)
	guestfsStateCleaned := guestfs.CleanStaleState()
	staleProvisionCleaned := op.Services.Cache.CleanStaleProvisionMounts(ctx, dryRun)

	result := map[string]any{
		"service_binaries":       serviceBinariesCleaned,
		"appliance":              appliancePruned,
		"warm_images":            warmPruned,
		"guestfs_state":          guestfsStateCleaned,
		"stale_provision_mounts": staleProvisionCleaned,
	}

	return result, nil
}

// CachePruneAll performs complete cache prune across all resource types.
// Matches Python's CacheOperation.prune_all() exactly.
// Returns *model.PruneAllResult matching Python's PruneAllResult dataclass.
func (op *Operation) CachePruneAll(ctx context.Context, dryRun bool, includeAll bool) (*model.PruneAllResult, error) {
	// Python: HostPrivilegeHelper.check_privileges("/usr/sbin/ip", "prune all cache resources")
	if err := system.CheckPrivileges("/usr/sbin/ip", "prune all cache resources"); err != nil {
		return nil, errs.WrapMsg(
			errs.CodePrivilegeRequired,
			fmt.Sprintf("Privilege check failed: %v", err),
			err,
			errs.WithClass(errs.ClassNeedsInteraction),
		)
	}

	// Detect running VMs (Python: Repository(db).list_all() then check status)
	hadRunningVMs := false
	if op.Repos.VM != nil {
		vms, err := op.Repos.VM.ListAll(ctx)
		if err == nil {
			for _, v := range vms {
				if v.Status == model.VMStatusRunning || v.Status == model.VMStatusStarting {
					hadRunningVMs = true
					break
				}
			}
		}
	}

	// Prune each resource type — collect IDs from results.
	var prunedIDs []string
	var failedIDs []string

	// CachePruneVMs still returns *errs.OperationResult (not yet refactored)
	pruneVMsResult := op.CachePruneVMs(ctx, dryRun, includeAll)
	if pruneVMsResult != nil && pruneVMsResult.IsOK() && pruneVMsResult.Item != nil {
		if items, ok := pruneVMsResult.Item.([]string); ok {
			prunedIDs = append(prunedIDs, items...)
		}
	}

	netIDs, err := op.CachePruneNetworks(ctx, dryRun, includeAll)
	if err == nil {
		prunedIDs = append(prunedIDs, netIDs...)
	}
	imgIDs, err := op.CachePruneImages(ctx, dryRun, includeAll)
	if err == nil {
		prunedIDs = append(prunedIDs, imgIDs...)
	}
	kernelIDs, err := op.CachePruneKernels(ctx, dryRun, includeAll)
	if err == nil {
		prunedIDs = append(prunedIDs, kernelIDs...)
	}
	binIDs, err := op.CachePruneBinaries(ctx, dryRun, includeAll)
	if err == nil {
		prunedIDs = append(prunedIDs, binIDs...)
	}

	miscMap, miscErr := op.CachePruneMisc(ctx, dryRun)
	if miscErr == nil {
		if infra.IsTrue(miscMap["appliance"]) {
			prunedIDs = append(prunedIDs, "appliance")
		}
		if infra.IsTrue(miscMap["warm_images"]) {
			prunedIDs = append(prunedIDs, "warm_images")
		}
		if infra.IsTrue(miscMap["guestfs_state"]) {
			prunedIDs = append(prunedIDs, "guestfs_state")
		}
		if infra.IsTrue(miscMap["stale_provision_mounts"]) {
			prunedIDs = append(prunedIDs, "stale_provision_mounts")
		}
	}

	// Use proper PruneAllResult struct matching Python's PruneAllResult dataclass
	result := &model.PruneAllResult{
		PrunedIDs:     prunedIDs,
		FailedIDs:     failedIDs,
		HadRunningVMs: hadRunningVMs,
	}

	return result, nil
}

// CacheClean performs complete cache clean.
// Matches Python's CacheOperation.clean() exactly.
// Returns *model.CleanResult matching Python's CleanResult dataclass.
func (op *Operation) CacheClean(ctx context.Context, dryRun bool) (*model.CleanResult, error) {
	// Step 1: Prune all cached resources (Python: CacheOperation.prune_all(dry_run=dry_run, include_all=True))
	pruneResult, pruneErr := op.CachePruneAll(ctx, dryRun, true)

	// Step 2: Abort if any VMs failed to prune or orphan processes remain post-prune.
	// (Python: checks prune_result.failed_ids and CacheService.scan_orphan_processes())
	if !dryRun && pruneErr == nil && pruneResult != nil {
		failedIDs := pruneResult.FailedIDs
		orphanProcesses := op.Services.Cache.ScanOrphanProcesses(ctx)

		if len(failedIDs) > 0 || len(orphanProcesses) > 0 {
			var messages []string
			if len(failedIDs) > 0 {
				messages = append(messages, fmt.Sprintf("Failed to remove %d VM(s): %s",
					len(failedIDs), strings.Join(failedIDs, ", ")))
			}
			if len(orphanProcesses) > 0 {
				pids := make([]string, 0, len(orphanProcesses))
				commSet := make(map[string]struct{})
				for _, p := range orphanProcesses {
					if pid, ok := p["pid"]; ok {
						pids = append(pids, fmt.Sprintf("%v", pid))
					}
					if comm, ok := p["comm"]; ok {
						commSet[fmt.Sprintf("%v", comm)] = struct{}{}
					}
				}
				names := make([]string, 0, len(commSet))
				for n := range commSet {
					names = append(names, n)
				}
				sort.Strings(names)
				messages = append(messages, fmt.Sprintf(
					"Orphan process(es) still running (PID(s) %s: %s). Kill them manually and re-run ``mvm clean``.",
					strings.Join(pids, ", "), strings.Join(names, ", ")))
			}

			return &model.CleanResult{
				PruneResult:     *pruneResult,
				CacheDirRemoved: false,
				CacheDir:        op.CacheDir,
			}, errs.New(errs.CodeCacheCleanFailed, strings.Join(messages, "; "))
		}
	}

	// Step 3: Clean host networking (while DB still exists in cache dir)
	// Python: HostOperation.clean(cache_dir) — unconditional call (hostOp is required constructor param)
	if !dryRun {
		_, _ = op.HostClean(ctx)
	}

	// Step 4: Remove the cache directory itself
	// Python: shutil.rmtree(cache_dir)
	cacheDirRemoved := false
	if _, err := os.Stat(op.CacheDir); err == nil {
		if !dryRun {
			os.RemoveAll(op.CacheDir)
		}
		cacheDirRemoved = true
	}

	// Build CleanResult (Python: CleanResult(prune_result=..., cache_dir_removed=..., cache_dir=...))
	if pruneResult == nil {
		pruneResult = &model.PruneAllResult{
			PrunedIDs:     []string{},
			FailedIDs:     []string{},
			HadRunningVMs: false,
		}
	}

	return &model.CleanResult{
		PruneResult:     *pruneResult,
		CacheDirRemoved: cacheDirRemoved,
		CacheDir:        op.CacheDir,
	}, nil
}
